### PR TITLE
DOC: 1.12.0 release notes

### PR DESCRIPTION
**Pull request by tylerjereddy**
* draft release notes for SciPy `1.12.0`, including transcription of the wiki version of the release notes

TODO:

- [x] manually go through the `95` merged PRs with `1.12.0` milestone and `enchancement` label, to fill in several parts of the release notes that were not filled in on the [wiki](https://github.com/scipy/scipy/wiki/*Release-Note-Entries-for-SciPy-1.12.0-(asterisk-makes-this-appear-at-the-top)) (this is normal--`stats` folks usually do a good job of filling in the wiki though)
- [x] fix formatting/build errors in the new release notes; nested bullet lists, indentation, the usual...
- [ ] `.mailmap` fixups to get a cleaner duplication/alias-free author list
- [x] deal with the `20` open PRs with `1.12.0` milestone, update the release notes again if/as needed (many of these will probably be milestone bumps on stagnant or difficult PRs)
- [ ] (after merging this): target the wheel builds back to released NumPy (not `2.x`), a faster-than-normal `1.13.0` is planned for NumPy `2.0` support a few months after this `1.12.0` release
- [ ] decide how much I/we care about having the best possible NumPy `main` support before we branch--I've put substantial effort into dealing with gh-19605, but strictly speaking, I don't think we need to do that, and things could still shift a bit anyway
- [x] decide on which new features/improvements qualify as the release highlights
- [x] add some sensible wording for "in progress, but cool/please test me" type features like array API support in some modules
- [x] progressively turn the CI back on here

[skip cirrus] [skip circle] [skip ci]

### CODE REVIEW

**Comment by lucascolley** [comment]
Some mailmap fixups at tylerjereddy/scipy#75

**Comment by mdhaber** [comment]
@tylerjereddy I hadn't filled in the wiki yet. Can I still do that or do you want changes to be made here?

**Comment by tylerjereddy** [comment]
> I hadn't taken a pass on the release notes for a few months. Can I still do that or do you want changes to be made here?

You can edit the wiki if you want, I'll check it again a few times anyway

**Comment by ev-br** [comment]
For scipy.interpolate:

- New class `NdBSpline` represents tensor-product splines in N dimensions. This class only known how to evaluate a tensor product given coefficients and knot vectors. This way it generalizes BSpline for 1D data to N-D, and parallels NdPPoly (which represents N-D tensor product polynomials). Evaluations exploit the localized nature of b-splines.

- `NeastNDInterpolator.__call__` accepts `**query_options`, which are passed through to the `KDTree.query` call to find nearest neighbors. This allows, for instance, to limit the neighbor search distance and parallelize the query using the `workers`  keyword. 

- `BarycentricInterpolator` now allows computing the derivatives.

**Comment by j-bowhay** [comment]
Probably worth mentioning #18279 in the deprecation section. Error messages have been made clearer for things that don't exist in the public namespace and warnings sharpened for private attributes that are not supposed to be imported at all

File: doc/source/release/1.12.0-notes.rst [hunk 1]
```diff
@@ -49,42 +59,83 @@ New features
...
+- `scipy.stats.anderson_ksamp` now supports a permutation version of the test
+  using the ``method`` parameter.
+- `scipy.stats.halcauchy.fit`, `scipy.stats.halflogistic.fit`, and
+  `scipy.stats.halfnorm.fit` are faster and more accurate.
+- `scipy.stats.beta.entropy` accuracy has been improved for extreme values of
+  distribution parameters.
+- The accuracy of ``sf`` and/or ``isf`` methods have been improved for
+  several distributions: `scipy.stats.burr`, `scipy.stats.hypsecant`,
+  `scipy.stats.kappa3`, `scipy.stats.loglaplace`, `scipy.stats.lognorm`,
+  `scipy.stats.lomax`, and `scipy.stats.pareto`.
```
Comment (mdhaber) [inline doc/source/release/1.12.0-notes.rst:138]:
```suggestion
  `scipy.stats.lomax`, and `scipy.stats.pareto`.
- The following functions now support parameters ``axis``, ``nan_policy``, and ``keep_dims``: `scipy.stats.entropy`, `scipy.stats.differential_entropy`, `scipy.stats.variation`, `scipy.stats.ansari`, `scipy.stats.bartlett`, `scipy.stats.levene`, `scipy.stats.fligner`, `scipy.stats.cirmean, `scipy.stats.circvar`, `scipy.stats.circstd`, `scipy.stats.tmean`, `scipy.stats.tvar`, `scipy.stats.tstd`, `scipy.stats.tmin`, `scipy.stats.tmax`, and `scipy.stats.tsem`.
- The ``logpdf`` and ``fit`` methods of `scipy.stats.skewnorm` have been improved.
- The beta negative binomial distribution is implemented as `scipy.stats.betanbinom`.
- The speed of `scipy.stats.invwishart` ``rvs`` and ``logpdf`` have been improved.
- A source of intermediate overflow in `scipy.stats.boxcox_normmax` with ``method='mle'`` has been eliminated, and the returned value of ``lmbda`` is constrained such that the transformed data will not overflow.
- `scipy.stats.nakagami.stats` is more accurate and reliable.
- A source of intermediate overflow in `scipy.norminvgauss.pdf` has been eliminated.
```

File: doc/source/release/1.12.0-notes.rst [hunk 2]
```diff
@@ -49,42 +59,83 @@ New features
...
--------------------------
-
-
-Other
------
+- masked array, axis tuple, and nan policy support was added to the trimmed
+  statistics functions: ``stats.tmean``, ``stats.tvar``, ``stats.tstd``,
+  ``stats.tmin``, ``stats.tmax``, and ``stats.tsem``.
+- ``axis`` and ``nan_policy`` support were added to `scipy.stats.levene` and
+  `scipy.stats.fligner`
```
Comment (mdhaber) [inline doc/source/release/1.12.0-notes.rst:112]: See combination of these in my other comment.
```suggestion
```

File: doc/source/release/1.12.0-notes.rst [hunk 3]
```diff
@@ -35,8 +35,30 @@ New features
 ============================
```
Comment (lucascolley) [inline doc/source/release/1.12.0-notes.rst:35]:
```suggestion
============================
- Experimental support added for the array API standard; PyTorch tensors,
  CuPy arrays and array API compatible array libraries are now accepted
  (GPU support is limited to functions with pure Python implementations).
  CPU arrays which can be converted to and from NumPy are supported
  module-wide and returned arrays will match the input type.
  This behaviour is enabled by setting the ``SCIPY_ARRAY_API`` environment
  variable before importing ``scipy``. This experimental support is still
  under development and likely to contain bugs - testing is very welcome.
```

File: doc/source/release/1.12.0-notes.rst [hunk 4]
```diff
@@ -35,8 +35,30 @@ New features
...
+  part of the ``fft`` array API standard extension module, as well as the 
+  Fast Hankel Transforms and the basic FFTs which are not in the extension
+  module, now accept PyTorch tensors, CuPy arrays and array API compatible
+  array libraries. CPU arrays which can be converted to and from NumPy arrays
+  are supported module-wide and returned arrays will match the input type.
+  This behaviour is enabled by setting the ``SCIPY_ARRAY_API`` environment
+  variable before importing ``scipy``. This experimental support is still under
+  development and likely to contain bugs - testing is very welcome.
+
 `scipy.interpolate` improvements
```
Comment (mdhaber) [inline doc/source/release/1.12.0-notes.rst:50]:
```suggestion
  development and likely to contain bugs - testing is very welcome.

`scipy.integrate` improvements
==============================
- Added `scipy.integrate.cumulative_simpson` for cumulative quadrature
  from data using Simpson's 1/3 rule.


`scipy.interpolate` improvements
```

**Comment by lucascolley** [comment]
Feel free to modify the `fft` and `cluster` array API parts to match what Matt has written for `special` if anything looks like an improvement.

File: doc/source/release/1.12.0-notes.rst [hunk 5]
```diff
@@ -49,42 +84,112 @@ New features
...
 `scipy.special` improvements
 ============================
+- Added `scipy.special.stirling2` for computation of Stirling numbers of the
+  second kind. Both exact calculation and an asymptotic approximation
+  (the default) are supported via ``exact=True`` and ``exact=False`` (the
+  default) respectively.
+- Experimental support added for alternative backends: functions
+  `scipy.special.log_ndtr`, `scipy.special.ndtr`, `scipy.special.ndtri`, 
+  `scipy.special.erf, `scipy.special.erfc`, `scipy.special.i0`, 
+  `scipy.special.i0e`, `scipy.special.i1, `scipy.special.i1e`,
```
Comment (lucascolley) [inline doc/source/release/1.12.0-notes.rst:147]:
```suggestion
  `scipy.special.erf`, `scipy.special.erfc`, `scipy.special.i0`, 
  `scipy.special.i0e`, `scipy.special.i1`, `scipy.special.i1e`, 
```

**Comment by tylerjereddy** [comment]
Obviously I'm a bit behind schedule, but I don't think we're too bad off, and folks have been helping a lot. So yeah, probably a little bit of a branching delay because end of year stuff made me a little busier, but not too worried I don't think.

File: doc/source/release/1.12.0-notes.rst [hunk 6]
```diff
@@ -49,42 +84,118 @@ New features
...
 ============================
+- Added `scipy.special.stirling2` for computation of Stirling numbers of the
+  second kind. Both exact calculation and an asymptotic approximation
+  (the default) are supported via ``exact=True`` and ``exact=False`` (the
+  default) respectively.
+- Experimental support added for alternative backends: functions
+  `scipy.special.log_ndtr`, `scipy.special.ndtr`, `scipy.special.ndtri`, 
+  `scipy.special.erf, `scipy.special.erfc`, `scipy.special.i0`, 
+  `scipy.special.i0e`, `scipy.special.i1, `scipy.special.i1e`, 
+  `scipy.special.gammaln`, `scipy.special.gammainc`, `scipy.special.gammaincc`,
```
Comment (dschmitz89) [inline doc/source/release/1.12.0-notes.rst:154]:
```suggestion
  `scipy.special.gammaln`, `scipy.special.gammainc`, `scipy.special.gammaincc`,
```

File: doc/source/release/1.12.0-notes.rst [hunk 7]
```diff
@@ -49,42 +84,118 @@ New features
...
+- Performance improvements have been added to ``distance.hamming`` and
+  ``distance.correlation``.
+- Improved performance of ``SphericalVoronoi`` ``sort_vertices_of_regions``.
 
 `scipy.special` improvements
 ============================
+- Added `scipy.special.stirling2` for computation of Stirling numbers of the
+  second kind. Both exact calculation and an asymptotic approximation
+  (the default) are supported via ``exact=True`` and ``exact=False`` (the
+  default) respectively.
```
Comment (dschmitz89) [inline doc/source/release/1.12.0-notes.rst:149]:
```suggestion
  default) respectively.
-  Added `scipy.special.betaincc` for computation of the complementary incomplete Beta function and `scipy.special.betainccinv` for computation of its inverse.
- Improved precision of `scipy.special.betainc` and `scipy.special.betaincinv`
```

PR: https://github.com/scipy/scipy/pull/17697

File: doc/source/release/1.12.0-notes.rst [hunk 8]
```diff
@@ -49,42 +84,118 @@ New features
...
+  circumstances. The new parameter is ``hess_inv0``.
+- ``minimize`` methods ``CG``, ``Newton-CG``, and ``BFGS`` now accept parameters
+  ``c1`` and ``c2``, allowing specification of the Armijo and curvature rule
+  parameters, respectively.
+- ``curve_fit`` performance has improved due to more efficient memoization
+  of the callable function.
 
 `scipy.signal` improvements
 ===========================
-
```
Comment (DietBru) [inline doc/source/release/1.12.0-notes.rst:56]: I think, due to its considerable length alone 😁, PR #17408 deserves a mention.  A possible wording could look like this:

```ReST
- The class ``ShortTimeFFT`` provides a new more versatile implementation of the
  short-time Fourier transform (STFT), its inverse (ISTFT) as well as the (cross-)
  spectrogram. It utilizes an improved algorithm for calculating the ISTFT based on
  dual windows and provides more fine-grained control of the parametrization especially
  in regard to scaling and phase-shift. Functionality was implemented to ease
  working with signal and STFT chunks. A section has been added to the "SciPy User Guide"
  providing algorithmic details. The functions ``stft``, ``istft`` and ``spectrogram``
  have been marked as legacy.
```

File: .mailmap [hunk 9]
```diff
@@ -325,6 +330,7 @@ Lindsey Hiltner <lindsey.hiltner@gmail.com> L. Hiltner <lhilt@users.noreply.gith
 Lindsey Hiltner <lindsey.hiltner@gmail.com> L Hiltner <lhilt@users.noreply.github.com>
 Lijun Wang <szcfweiya@gmail.com> szcf-weiya <szcfweiya@gmail.com>
 Lorenzo Luengo <> loluengo <>
+Lucas Colley <lucas.colley8@gmail.com> lucascolley <lucas.colley8@gmail.com>
```
Comment (lucascolley) [inline .mailmap:333]:
```suggestion
Lucas Colley <lucas.colley8@gmail.com> lucascolley <51488791+lucascolley@users.noreply.github.com>
```

**Comment by tylerjereddy** [comment]
I checked off the manual comb-through of the enhancement PRs and re-checking of the wiki. I believe there have been some merges since I started doing my pass, but anyway it took me a few days because of various interruptions.

I'll think a bit more about the highlights next, then fixing up the formatting/doc build.

**Comment by tylerjereddy** [comment]
Doc build is passing locally--let's see if CI agrees (minus cirrus for now).

File: doc/source/release/1.12.0-notes.rst [hunk 10]
```diff
@@ -33,73 +47,243 @@ New features
...
+- `scipy.stats.geometric_discrepancy` adds geometric/topological discrepancy
+  metrics for random samples.
+- `scipy.stats.multivariate_normal` now has a ``fit`` method for fitting
+  distribution parameters to data via maximum likelihood estimation.
+- `scipy.stats.bws_test` performs the Baumgartner-Weiss-Schindler test of
+  whether two-samples were drawn from the same distribution.
+- `scipy.stats.jf_skew_t` implements the Jones and Faddy skew-t distribution.
+- `scipy.stats.anderson_ksamp` now supports a permutation version of the test
+  using the ``method`` parameter.
+- The ``fit`` methods of `scipy.stats.halcauchy`, `scipy.stats.halflogistic`, and
```
Comment (dschmitz89) [inline doc/source/release/1.12.0-notes.rst:218]:
```suggestion
- The ``fit`` methods of `scipy.stats.halfcauchy`, `scipy.stats.halflogistic`, and
```

File: doc/source/release/1.12.0-notes.rst [hunk 11]
```diff
@@ -33,73 +47,243 @@ New features
...
+  `scipy.stats.halfnorm` are faster and more accurate.
+- `scipy.stats.beta` ``entropy`` accuracy has been improved for extreme values of
```
Comment (dschmitz89) [inline doc/source/release/1.12.0-notes.rst:220]:
```suggestion
- ``scipy.stats.beta.entropy`` accuracy has been improved for extreme values of
```

File: doc/source/release/1.12.0-notes.rst [hunk 12]
```diff
@@ -33,73 +47,243 @@ New features
...
+- The ``logpdf`` and ``fit`` methods of `scipy.stats.skewnorm` have been improved.
+- The beta negative binomial distribution is implemented as `scipy.stats.betanbinom`.
+- The speed of `scipy.stats.invwishart` ``rvs`` and ``logpdf`` have been improved.
+- A source of intermediate overflow in `scipy.stats.boxcox_normmax` with ``method='mle'`` has been eliminated, and the returned value of ``lmbda`` is constrained such that the transformed data will not overflow.
+- `scipy.stats.nakagami` ``stats`` is more accurate and reliable.
+- A source of intermediate overflow in `scipy.norminvgauss.pdf` has been eliminated.
+- Added support for masked arrays to ``stats.circmean``, ``stats.circvar``,
+  ``stats.circstd``, and ``stats.entropy``.
+- ``gausshyper`` ``cdf`` and ``rvs`` methods have improved performance.
```
Comment (dschmitz89) [inline doc/source/release/1.12.0-notes.rst:236]:
```suggestion
```

The PR https://github.com/scipy/scipy/pull/18799 this refers to  is difficult to decipher: in the end it was basically a simplifcation of a formula with no notable impact for users.

**Review by dschmitz89 — CHANGES_REQUESTED** [review]
A few minor things, mostly typos.

**Comment by tylerjereddy** [comment]
I still see blemishes in the rendered version of the release notes for `1.12.0`, but CI is green (including ARM architectures and doc build) and I'm behind schedule on this release cycle so I'll proceed with (squash) merge and the release notes can continue to be improved during the release process (this happens pretty much every time to be honest).

**Pull request merged**